### PR TITLE
Add ability to merge using multiple columns in JOIN condition

### LINKED ISSUE #5657: DataFrame Merge Multiple Columns
The `Merge` function can only take two strings for column names on the merge. It would be great if we can merge across multiple columns.

As a suggestion, maybe something like the pandas merge statement. Below is from the merge statement documentation in pandas:

on: label or list
Column or index level names to join on. These must be found in both DataFrames. If on is None and not merging on indexes then this defaults to the intersection of the columns in both DataFrames.

left_on: label or list, or array-like
Column or index level names to join on in the left DataFrame. Can also be an array or list of arrays of the length of the left DataFrame. These arrays are treated as if they are columns.

right_on: label or list, or array-like
Column or index level names to join on in the right DataFrame. Can also be an array or list of arrays of the length of the right DataFrame. These arrays are treated as if they are columns.

I was not able to find another similar issue. If there is an agreement on the method signatures, I am happy to put together a PR for this.

### PR DESCRIPTION
**Pull request by asmirnov82**
Fixes #5657

PR contains 3 main changes (these changes were done independently, step by step):

1) Avoid code duplication in Merge implementation (commit - dc8daf0ffbcf2a26757daee988c2ddbeebf1e646). To achive that I created new static function, that merges 2 dataframes (retained and supplementary). All merge implementations (Left. right, inner and outer) uses this function to perform correct merge. 

2) Change templated (generic) DataFrame Merge<TKey> (...) function to general (not templated) DataFrame Merge(...)
Commit a4735ce82a7860d5c355472b88fc0b05cd54d50b. To achive the goal I added method to find  element entrance of one column into another inside the column (column is already templated). Base column implementation provides 2 methods: one is public not templated and abstract (used like an interface).

`public abstract Dictionary<long, ICollection<long>> GetGroupedOccurrences(DataFrameColumn other, out HashSet<long> otherColumnNullIndices);`

Second provides real implementation and is templated:

`protected Dictionary<long, ICollection<long>> GetGroupedOccurrences<TKey>(DataFrameColumn other, out HashSet<long> otherColumnNullIndices)`

Each column implements abstract method by calling protected method providing correct Type as a generic

This step was neccessary to allow group by multiple columns, that may have different data types

3) Actual implementation of merging by multiple columns (all other commits), by finding occurrences on each column from the JOIN list and finding intersection of these occurrences

### CODE REVIEW

File: src/Microsoft.Data.Analysis/DataFrameColumn.cs [hunk 1]
```diff
@@ -210,6 +210,44 @@ public virtual DataFrameColumn Sort(bool ascending = true)
...
+            // First hash other column   
+            Dictionary<TKey, ICollection<long>> multimap = other.GroupColumnValues<TKey>(out otherColumnNullIndices);
+
+            var ret = new Dictionary<long, ICollection<long>>();
+
+            //For each value in this column find rows from other column with equal value
+            for (int i = 0; i < this.Length; i++)
+            {
+                var value = this[i];
+                if (value != null && multimap.TryGetValue((TKey)value, out ICollection<long> otherRowIndices))
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrameColumn.cs:242]: Q for myself (I'm half way through the changes): What happens to null values here?

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs [hunk 2]
```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
-                            }
-                        }
-                        else
-                        {
-                            leftRowIndices.Append(null);
-                            rightRowIndices.Append(i);
-                        }
-                    }
-                    else
+                    //only rows with occurences from previose step should go for futher processing
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.Join.cs:198]: Nit: previose -> previous

File: src/Microsoft.Data.Analysis/DataFrameColumn.cs (hunk 1 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrameColumn.cs:242]: Answer: It's handled outside this function in the `Merge` method. Perhaps 1 thing to consider here are the names `GetGroupedOccurrences` vs `GroupColumnValues`. `GetGroupedOccurrences` works on "other" DataFrameColumns while `GroupColumnValues` works on the current `DataFrameColumn`. However, I'm not sure that distinction comes across from the API names. I'll see if I can come up with better names.

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs [hunk 3]
```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
-                            leftRowIndices.Append(thisColumnNullIndex);
-                            rightRowIndices.Append(i);
-                        }
+                        //store reverse mapping to restore original dataframe indices from indices in shrinked row
+                        var originalIndex = shrinkedRetainedIndices[i];
+                        newRetainedIndicesReverseMapping.Add(i, originalIndex);
                     }
+
+                    retainedIndicesReverseMapping = newRetainedIndicesReverseMapping;
+                    shrinkedRetainedColumn = shrinkedRetainedColumn.Clone(new Int64DataFrameColumn("Indices", shrinkedRetainedIndices));
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.Join.cs:212]: Nice!

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs [hunk 4]
```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
-                DataFrameColumn otherColumn = ReferenceEquals(hashColumn, Columns[leftJoinColumn]) ? other.Columns[rightJoinColumn] : Columns[leftJoinColumn];
-                Dictionary<TKey, ICollection<long>> multimap = hashColumn.GroupColumnValues<TKey>(out HashSet<long> smallerDataFrameColumnNullIndices);
+                //Convert indices from in key from local (shrinked row) to indices in original dataframe
+                if (retainedIndicesReverseMapping != null)
+                    newOccurrences = newOccurrences.ToDictionary(kvp => retainedIndicesReverseMapping[kvp.Key], kvp => kvp.Value);
 
-                for (long i = 0; i < otherColumn.Length; i++)
+                supplementaryJoinColumnsNullIndices.UnionWith(supplementaryColumnNullIndices);
+                
+                // shrink join result on current column by previouse join columns (if any)
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.Join.cs:226]: Nit: previouse -> previous

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs [hunk 5]
```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
+                // (we have to remove occurrences that doesn't exist in previouse columns, because JOIN happens only if ALL left and right columns in JOIN are matched)
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.Join.cs:227]: Nit: ditto

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs [hunk 6]
```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
-                        }
-                        else
-                        {
-                            leftRowIndices.Append(null);
-                            rightRowIndices.Append(i);
-                        }
-                    }
-                    else
+                    //only rows with occurences from previose step should go for futher processing
+                    var shrinkedRetainedIndices = occurrences.Keys.ToArray();
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.Join.cs:199]: Q for next round of review: Is the call to `ToArray` necessary? I'm thinking about how performant this method might be.

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs [hunk 7]
```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
+                //Find occurrenses on current step (join column)
+                var newOccurrences = shrinkedRetainedColumn.GetGroupedOccurrences(supplementaryColumn, out HashSet<long> supplementaryColumnNullIndices);
 
-                bool leftColumnIsSmaller = leftRowCount <= rightRowCount;
-                DataFrameColumn hashColumn = leftColumnIsSmaller ? Columns[leftJoinColumn] : other.Columns[rightJoinColumn];
-                DataFrameColumn otherColumn = ReferenceEquals(hashColumn, Columns[leftJoinColumn]) ? other.Columns[rightJoinColumn] : Columns[leftJoinColumn];
-                Dictionary<TKey, ICollection<long>> multimap = hashColumn.GroupColumnValues<TKey>(out HashSet<long> smallerDataFrameColumnNullIndices);
+                //Convert indices from in key from local (shrinked row) to indices in original dataframe
+                if (retainedIndicesReverseMapping != null)
+                    newOccurrences = newOccurrences.ToDictionary(kvp => retainedIndicesReverseMapping[kvp.Key], kvp => kvp.Value);
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.Join.cs:222]: Going from array->dictionary here. Another place to consider performance.

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs [hunk 8]
```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
-                Int64DataFrameColumn thisColumnNullIndices = new Int64DataFrameColumn("ThisColumnNullIndices");
+                occurrences = newOccurrences;
+            }
+            
+            retainedRowIndices = new Int64DataFrameColumn("RetainedIndices");
+            supplementaryRowIndices = new Int64DataFrameColumn("SupplementaryIndices");
 
-                for (long i = 0; i < thisColumn.Length; i++)
+            //Perform Merging 
+            var retainJoinColumns = retainedJoinColumnNames.Select(name => retainedDataFrame.Columns[name]).ToArray();
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.Join.cs:250]: Is `retainJoinColumns` here the same as just writing a for loop over `retainedJoinColumnNames`? If so, we should get rid of LINQ here to be performant.

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs [hunk 9]
```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
...
-                            }
-                        }
-                    }
-                    else
+                    var shrinkedOccurences = new Dictionary<long, ICollection<long>>();
+
+                    foreach (var kvp in newOccurrences)
                     {
-                        foreach (long nullIndex in smallerDataFrameColumnNullIndices)
+                        var newValue = kvp.Value.Where(i => occurrences[kvp.Key].Contains(i)).ToArray();
```
Comment (pgovind) [inline src/Microsoft.Data.Analysis/DataFrame.Join.cs:234]: Q for next round of review: Is LINQ necessary here?

**Review by pgovind — COMMENTED** [review]
I left a few comments. They are mostly for myself right now. Please don't answer them yet. I will do another round of reviewing these changes tomorrow.

Overall this PR is looking quite good. I like the changes, great job! Apologies for the late review here. I was super busy with .NET 6 stuff that I didn't have time to focus on DataFrame for 2 months. I'm relatively free all of September though, so I can review changes as they come in for DataFrame.

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs (hunk 2 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/DataFrame.Join.cs:198]: fixed

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs (hunk 4 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/DataFrame.Join.cs]: fixed

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs (hunk 5 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/DataFrame.Join.cs]: fixed

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs (hunk 7 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/DataFrame.Join.cs:222]: we are going from dictionary to dictionary, the idea is that we have to convert indices from _shrinkedRetainedColumn_ (to wich we apply GetGroupedOccurrences) to the indices for original data frame. Possible solution is to provide _retainedIndicesReverseMapping_ inside 
_protected Dictionary<long, ICollection<long>> GetGroupedOccurrences<TKey>(DataFrameColumn other, out HashSet<long> otherColumnNullIndices)_ as optional parameter to create correct indices just in place. What do you think, should we do that?

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs (hunk 8 above)
Comment (asmirnov82) [reply src/Microsoft.Data.Analysis/DataFrame.Join.cs:250]: `retainJoinColumns` is a collection of columns, that are used for joining. It is used in `IsAnyNullValueInColumns` method for each row in the loop to get if there is any null value at least of one of the column, that is used as part of the key used for joining

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs (hunk 6 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.Join.cs:199]: This is ok. If users are merging with multiple columns, I think it;s ok to pay this price.

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs (hunk 7 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.Join.cs:222]: Yup, you're right, it's dictionary -> dictionary. The diff on GH made it hard to find out the type of newOccurences. I think this is ok. It's more readable IMO

File: src/Microsoft.Data.Analysis/DataFrame.Join.cs (hunk 9 above)
Comment (pgovind) [reply src/Microsoft.Data.Analysis/DataFrame.Join.cs:234]: @asmirnov82 : I think you can get rid of the ToArray here right? `newValue` will then be an `IEnumerable<long>` and you can still call `newValue.Any()` in the next line.

**Review by pgovind — APPROVED** [review]
Ok, done with the 2nd review. LGTM. Thanks for the work here @asmirnov82. This is a great addition to DataFrame!

**Comment by pgovind** [comment]
I looked at the test failures and it looks unrelated. Do you mind rebasing on the latest main branch? That might fix the CI failures and let us merge this PR

**Comment by asmirnov82** [comment]
> I looked at the test failures and it looks unrelated. Do you mind rebasing on the latest main branch? That might fix the CI failures and let us merge this PR

I merged latest into my repo, this didn't help. Several tests failed, due to infrastructure issue, for example one with unability to load nuget package due to connection timeout (".dotnet\sdk\6.0.100-preview.3.21202.5\NuGet.targets(131,5): error : (NETCORE_ENGINEERING_TELEMETRY=Restore) Failed to download package 'runtime.native.System.4.3.0' from 'https://pkgs.dev.azure.com/dnceng/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_packaging/45bacae2-5efb-47c8-91e5-8ec20c22b4f8/nuget/v3/flat2/runtime.native.system/4.3.0/runtime.native.system.4.3.0.nupkg'.
A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond. (c78vsblobprodcus322.vsblob.vsassets.io:443)
  A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond.)

**Comment by pgovind** [comment]
@michaelgsharp : I'm seeing this error in CI:

```
Starting test: Microsoft.ML.Tests.ImageTests.TestGreyscaleTransformImages
    Microsoft.ML.Tests.ImageTests.TestBackAndForthConversionWithoutAlphaNoInterleave [FAIL]
[31;1m[m[37m      System.ArgumentException : Parameter is not valid.
[m[30;1m      Stack Trace:
[m[37m           at System.Drawing.Bitmap.UnlockBits(BitmapData bitmapdata)
[m[37m        /__w/1/s/src/Microsoft.ML.ImageAnalytics/VectorToImageTransform.cs(421,0): at Microsoft.ML.Transforms.Image.VectorToImageConvertingTransformer.Mapper.<>c__DisplayClass5_0`1.<GetterFromType>b__0(Bitmap& dst)
[m[37m        /__w/1/s/test/Microsoft.ML.Tests/ImagesTests.cs(573,0): at Microsoft.ML.Tests.ImageTests.TestBackAndForthConversionWithoutAlphaNoInterleave()
[m[30;1m      Output:
[m[37m        Test TestBackAndForthConversionWithoutAlphaNoInterleave: aborted: passed
[mFinished test: Microsoft.ML.Tests.ImageTests.TestGreyscaleTransformImages with memory usage 172,474,368.00
Starting test: Microsoft.ML.Tests.ImageTests.TestBackAndForthConversionWithAlphaInterleave
Finished test: Microsoft.ML.Tests.ImageTests.TestBackAndForthConversionWithAlphaInterleave with memory usage 176,336,896.00
Starting test: Microsoft.ML.Tests.ImageTests.TestSaveImages
Finished test: Microsoft.ML.Tests.ImageTests.TestSaveImages with memory usage 177,635,328.00
[37m  Finished:    Microsoft.ML.Tests
```

Any idea if this is being tracked somewhere?

**Comment by michaelgsharp** [comment]
@pgovind its not yet, but I did bring this up yesterday. There are some tests that are flaky, and this is one of them. It doesn't happen always, and I believe its only with arm stuff right? It would be good to make an issue about flaky tests

**Comment by pgovind** [comment]
Unrelated CI failure. Merging

**Pull request merged**